pgjdbc/src/main/java/org/postgresql/util/LruCache.java

```diff
@@ -74,27 +76,73 @@ protected boolean removeEldestEntry(Map.Entry<Key, Value> eldest) {
     }
   }
 
-  private void evictValue(Value value) {
+  final void evictValue(Value value) {
```

Comment:
Why is it necessary for this method to be exposed to the package scope, and why is it necessary for it to be "final"? From an API perspective, it provides maximum flexibility to not expose private names, as it avoids people accidentally relying on them. You mentioned synthetic accessors... have you been able to measure a difference here? I don't think the code is better after you left it in this case - it seems better before.

Reply:
A default scoped method is required because limited map is calling it. If the code is defined as private, then the compiler will generate a default scoped wrapper for limited map to call through to this. What is downside of just making this default scoped?
I am accustomed to having this as a compiler warning and being expected to resolve is classes otherwise being modified. If this project prefers to rely on the synthetic accessors I will revert.

Reply:
Once we target jdk 11 the synthetic accessors are no longer required.
https://openjdk.java.net/jeps/181

Reply:
My personal preference, although I don't set the project-wide objectives and am just sharing my personal preference...

1. First ensure that the code is correct. It has the correct API (including visibility) and behaviour with the lowest complexity implementation that still meets all requirements. With practice, this should result in generally good code that is scalable and performant.
2. If necessary, or if it can be predicted as necessary, identify what the real performance of the system is, and if the performance is not acceptable for some reason, consider making compromises to simplicity to address the performance concerns.

Based upon the above general approach, I would not consider "do I want to avoid synthetic methods?" a day-to-day question. Only if a particular hotspot was found, and a detailed analysis determined that the synthetic methods were causing a real problem, would I choose to go back and try and re-work the solution to avoid synthetic methods, or consider an alternate implementation which might be both simple and more performance than the current one.

It's good to know that Java 11 may eliminate the need for the synthetic accessors. This is another major benefit of Java compared to other languages - that if we write the common in common ways, then common solutions which benefit everyone can be produced, vs every designer having to optimize on their own. Java can be a bit of a beast in terms of runtime overhead, so we may as well make the best use of it. No point in having a turbo-charged JIT, if we write the code the same way we would if it was C... :-)

Reply:
> I've noted that many highly "optimized" code bits in Java are so complex as to invalidate the JIT - for example, class methods that are too complex to inline. Java can be a bit of a beast in terms of runtime overhead, so we may as well make the best use of it.

This is a bit different than that. The compiler is telling us that the accessing of private scoped variables/methods from the inner class is adding complexity to the bytecode. The fact that the jvm access rules were changed in jdk 11 to permit this would seem to indicate there is a legitimate downside to the synthetic accessors. While it is possible that jvm can optimize it all away at runtime, that burden can be removed with the visibility change. At a conceptual level we are in agreement. I just do not view the change from private to default scope to be as consequential as you appear to.

Reply:
I think it would be best to follow the style of the project's code as of now, which appears to generate synthetic accessors. If that was to be avoided, then it should be decided project wide and in a dedicated issue for discussion. If it is a one-off then why should be communicated in the code, e.g. Guava includes `@VisibleForTesting` as a common enough occurrence in large code bases. 

Caffeine uses default scope both to avoid this extra byte code and to simplify tests that inspect the internal state. That is project wide and tests have to inspect internals in-depth due to being a data structure utility. I wouldn't carry that to application code or foreign code bases where it may not make as much sense to other developers. In this case where it is also a cache and test inspection might be desirable, I'd default to private but use `// @VisibleForTesting` if needed for testing (and commented out to avoid the dependency). I wouldn't worry about the synthetic byte code as far too small of an optimization given other costs and it may be optimized away by the JDK anyway.